pool/app/filter/entity.ml

```diff
@@ -297,85 +303,282 @@ module Key = struct
-    | ContainsNone [@printer print "contains_none"] [@name "contains_none"]
-    | ContainsAll [@printer print "contains_all"] [@name "contains_all"]
-  [@@deriving show { with_path = false }, eq, enum, yojson]
-
-  let input_type_to_operator =
-    let open Key in
-    function
-    | Key.Bool -> [ Equal; NotEqual ]
-    | Key.Date | Nr ->
-      [ Equal; NotEqual; Greater; GreaterEqual; Less; LessEqual ]
-    | Key.Languages _ -> [ Equal; NotEqual ]
-    | MultiSelect _ | QueryExperiments ->
-      [ ContainsAll; ContainsSome; ContainsNone ]
-    | Select _ -> [ Equal; NotEqual ]
-    | Str -> [ Equal; NotEqual; Like ]
+    | Equality of Equality.t
+    | Existence of Existence.t
+    | List of ListM.t
+    | Size of Size.t
+    | String of StringM.t
+  [@@deriving show { with_path = false }, eq, yojson, variants]
+
+  let show = function
+    | Equality o -> Equality.show o
+    | String o -> StringM.show o
+    | Size o -> Size.show o
+    | List o -> ListM.show o
+    | Existence o -> Existence.show o
+  ;;
+
+  let all_equality_operators = Equality.all >|= equality
+  let all_string_operators = StringM.all >|= string
+  let all_size_operators = Size.all >|= size
+  let all_list_operators = ListM.all >|= list
+  let all_existence_operators = Existence.all >|= existence
+
+  let all =
+    let open CCList in
+    all_equality_operators
+    @ all_string_operators
+    @ all_size_operators
+    @ all_list_operators
+    @ all_existence_operators
   ;;
 
-  let of_string m =
-    try Ok (`String m |> t_of_yojson) with
-    | _ -> Error Pool_common.Message.(Invalid Field.Operator)
+  let encode_operators =
+    let open CCOption.Infix in
+    [ Equality.(all |> CCList.map show, fun y -> y |> read >|= equality)
+    ; StringM.(all |> CCList.map show, fun y -> y |> read >|= string)
+    ; Size.(all |> CCList.map show, fun y -> y |> read >|= size)
+    ; ListM.(all |> CCList.map show, fun y -> y |> read >|= list)
+    ; Existence.(all |> CCList.map show, fun y -> y |> read >|= existence)
+    ]
+    |> CCList.flat_map (fun (operators, fnc) ->
+         CCList.map (fun operator -> operator, fnc) operators)
   ;;
 
   let of_yojson yojson =
-    try Ok (yojson |> read t_of_yojson) with
-    | _ -> Error Pool_common.Message.(Invalid Field.Operator)
+    let rec find_operator lst item =
+      match lst with
+      | [] -> None
+      | (operator, fnc) :: tl ->
+        if CCString.equal operator item
+        then fnc yojson
+        else find_operator tl item
+    in
+    (match yojson with
+     | `String str -> find_operator encode_operators str
+     | _ -> None)
+    |> CCOption.to_result Pool_common.Message.(Invalid Field.Operator)
   ;;
 
-  let to_yojson m = `String (m |> show)
+  let yojson_of_t m : Yojson.Safe.t =
+    (match m with
+     | Equality o -> Equality.show o
+     | String o -> StringM.show o
+     | Size o -> Size.show o
+     | List o -> ListM.show o
+     | Existence o -> Existence.show o)
+    |> fun str -> `String str
```

Comment:
I do not use the deriving yojson functions because they would produce something like `["Equality, ["equal"]"]`. The UI would have to produce JSON in this format as well.

Using the show functions inside the submodules removes the prefix.